src/extensions/struct-conn/index.ts

```diff
@@ -0,0 +1,365 @@
+        const propTests: Parameters<typeof struct.generator.atomGroups>[0] = {
+            'chain-test': core.rel.eq([struct.atomProperty.macromolecular.label_asym_id(), partner.asymId]),
+            'group-by': struct.atomProperty.core.operatorName(),
+        };
+        if (partner.seqId !== undefined) {
+            propTests['residue-test'] = core.rel.eq([struct.atomProperty.macromolecular.label_seq_id(), partner.seqId]);
+        } else if (partner.authSeqId !== undefined) { // for the case of water and carbohydrates (see 5elb, covale3 vs covale5)
+            propTests['residue-test'] = core.logic.and([
+                core.rel.eq([struct.atomProperty.macromolecular.auth_seq_id(), partner.authSeqId]),
+                core.rel.eq([struct.atomProperty.macromolecular.pdbx_PDB_ins_code(), partner.insCode]),
+            ]);
+        }
+        if (by === 'residues' && partner.altId !== '') {
+            propTests['atom-test'] = core.rel.eq([struct.atomProperty.macromolecular.label_alt_id(), partner.altId]);
+        }
+        if (by === 'atoms') {
+            propTests['atom-test'] = core.logic.and([
+                core.rel.eq([struct.atomProperty.macromolecular.label_atom_id(), partner.atomId]),
+                core.rel.eq([struct.atomProperty.macromolecular.label_alt_id(), partner.altId]),
+            ]);
+        }
+        partnerExpressions.push(struct.filter.first([struct.generator.atomGroups(propTests)]));
+    }
+    return struct.combinator.merge(partnerExpressions.map(e => struct.modifier.union([e])));
+}
+
+/** Create visuals for residues and atoms involved in a struct_conn and zoom on them */
+async function addStructConnInspection(plugin: PluginContext, structNode: StructNode, conn: StructConnRecord) {
+    if (!structNode) return;
+
+    const expressionByResidues = structConnExpression(conn, 'residues');
+    const expressionByAtoms = structConnExpression(conn, 'atoms');
+
+    const selectionByResidues = await plugin.build().to(structNode).apply(
+        StructureComponent,
+        { label: `${conn.id} (residues)`, type: { name: 'expression', params: expressionByResidues } },
+        { tags: [TAGS.RESIDUE_SEL] }
+    ).commit();
+    const selectionByAtom = await plugin.build().to(structNode).apply(
+        StructureComponent,
+        { label: `${conn.id} (atoms)`, type: { name: 'expression', params: expressionByAtoms } },
+        { tags: [TAGS.ATOM_SEL] }
+    ).commit();
+
+    if (selectionByResidues.isOk) {
+        await plugin.build().to(selectionByResidues).apply(
+            StructureRepresentation3D,
+            RESIDUES_VISUAL_PARAMS,
+            { tags: [TAGS.RESIDUE_REPR] }
+        ).commit();
+    }
+    if (selectionByAtom.isOk) {
+        const atomBalls = await plugin.build().to(selectionByAtom).apply(
+            StructureRepresentation3D,
+            ATOMS_VISUAL_PARAMS,
+            { tags: [TAGS.ATOM_REPR] }
+        ).commit();
```

Comment:
You should be able to do all of this in a single commit. If a selection is empty, the visual will be skipped (hopefully :)).

At the very least, the selections can be committed together and so can the the visuals.

Reply:
> If a selection is empty, the visual will be skipped (hopefully)
You're right, it will be skipped if it's in one commit (but it it's two commits and the selection is empty, the visual would apply to the whole structure!).
I was adding more visuals in the older version, that's why I split the commits.

Reply:
I will do this:
```
    const residueVis = await plugin.build().to(structNode).apply(StructureComponent,...).apply(StructureRepresentation3D, ...).commit();
    const atomVis = await plugin.build().to(structNode).apply(StructureComponent,...).apply(StructureRepresentation3D, ...).commit();
```

Reply:
I think that's even worse. If you are doing both, you can do:

```ts
const update = plugin.build()
const residueVis = update.to(structNode).apply(StructureComponent,...).apply(StructureRepresentation3D, ...).selector;
const atomVis = ....
await update.commit();
```

Reply:
> (but it it's two commits and the selection is empty, the visual would apply to the whole structure!).

This is actually a bug then.

Reply:
OK, somehow I was struggling with getting the atomVis node then (to zoom on it), but now I found a way though refs.

Reply:
Should be fixed now.